### PR TITLE
[dnf] Fix skip_broken, add test coverage

### LINKED ISSUE #73072: skip_broken in dnf module doesnt have any effect
##### SUMMARY

Trying to install a list of local packages via dnf module gives me conflicts on RHEL 8.2
I tried to add skip_broken flag but the conflicts error are still appear

```dnf --disablerepo=* localinstall -y *.rpm --skip-broken```
works well

Important notice: server should be without an outbound connection. (disconnected from internet)

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME

dnf

##### ANSIBLE VERSION

```paste below
ansible 2.10.4
  config file = /mnt/ansible/ansible.cfg
  configured module search path = ['/root/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/local/lib/python3.6/site-packages/ansible
  executable location = /usr/local/bin/ansible
  python version = 3.6.8 (default, Aug 18 2020, 08:33:21) [GCC 8.3.1 20191121 (Red Hat 8.3.1-5)]
```

##### CONFIGURATION

```paste below
DEFAULT_CALLBACK_WHITELIST(/mnt/ansible/ansible.cfg) = ['timer', 'profile_tasks']
DEFAULT_HASH_BEHAVIOUR(/mnt/ansible/ansible.cfg) = merge
DEFAULT_HOST_LIST(/mnt/ansible/ansible.cfg) = ['/mnt/ansible/inventory']
DEFAULT_LOAD_CALLBACK_PLUGINS(/mnt/ansible/ansible.cfg) = True
DEFAULT_LOG_PATH(/mnt/ansible/ansible.cfg) = /mnt/ansible/ansible.log
DEFAULT_REMOTE_USER(/mnt/ansible/ansible.cfg) = ec2-user
DEFAULT_STDOUT_CALLBACK(/mnt/ansible/ansible.cfg) = debug
DEFAULT_VERBOSITY(/mnt/ansible/ansible.cfg) = 0
HOST_KEY_CHECKING(/mnt/ansible/ansible.cfg) = False
INVENTORY_ENABLED(/mnt/ansible/ansible.cfg) = ['host_list', 'virtualbox', 'constructed', 'script', 'auto', 'yaml', 'ini', 'toml']
RETRY_FILES_ENABLED(/mnt/ansible/ansible.cfg) = False
```

##### OS / ENVIRONMENT

RHEL 8.2 running on Azure

##### STEPS TO REPRODUCE

1. on RHEL 8.3 - download nginx (for example) rpms and dependencies to a folder (dnf --downloadonly)
2. Copy all  to RHEL 8.2
3. try to install these local rpms via dnf module

```yaml
- name: INSTALL DEPENDENCIES ROLE | NGINX | INSTALL | Finding Nginx RPM files
  find:
    paths: "{{ local_repo_nginx_dir_path }}"
    patterns: "*.rpm"
  register: rpm_result

- set_fact:
    rpm_list: "{{ rpm_result.files | map(attribute='path') | list}}"

- name: INSTALL DEPENDENCIES ROLE | NGINX | INSTALL | Install all Nginx RPMs dependencies
  dnf:
    name: "{{ rpm_list }}"
    state: present
    disablerepo: "*"
    skip_broken: yes
```

##### EXPECTED RESULTS

nginx installation doesn't fail and playbook continue

##### ACTUAL RESULTS

playbook fails on dnf module

```paste below
MSG:

Depsolve Error occured:
 Problem 1: package crypto-policies-20200713-1.git51d1222.el8.noarch conflicts with openssh < 8.0p1-5 provided by openssh-8.0p1-4.el8_1.x86_64
  - conflicting requests
  - problem with installed package openssh-8.0p1-4.el8_1.x86_64
 Problem 2: cannot install both util-linux-2.32.1-24.el8.x86_64 and util-linux-2.32.1-22.el8.x86_64
  - package util-linux-user-2.32.1-22.el8.x86_64 requires util-linux = 2.32.1-22.el8, but none of the providers can be installed
  - conflicting requests
  - problem with installed package util-linux-user-2.32.1-22.el8.x86_64
```

### PR DESCRIPTION
**Pull request by relrod**
##### SUMMARY

Change:
- skip_broken was set in config but not actually used in calls to
  base.install()
- added a lot of test cases with specialized repo
- got rid of external (docker repo) nobest test cases since the
  specialized repo works well for those too
- Slight cleanup and adding comments in dnf module

Test Plan:
- ci_complete

Tickets:
- Fixes #73072

Signed-off-by: Rick Elrod <rick@elrod.me>

##### ISSUE TYPE

- Bugfix Pull Request
- Test Pull Request

##### COMPONENT NAME

dnf

### CODE REVIEW

File: lib/ansible/modules/dnf.py [hunk 1]
```diff
@@ -1079,9 +1081,13 @@ def ensure(self):
...
-                        # best effort causes to install the latest package
-                        # even if not previously installed
-                        self.base.conf.best = True
+                        # Previously we forced base.conf.best=True here.
+                        # However in 2.11+ there is a self.nobest option, so defer to that.
+                        # Note, however, that just because nobest isn't set, doesn't mean that
+                        # base.conf.best is actually true. We only force it false in
+                        # _configure_base(), we never set it to true, and it can default to false.
+                        # Thus, we still need to explicitly set it here.
+                        self.base.conf.best = not self.nobest
```
Comment (samdoran) [inline lib/ansible/modules/dnf.py:1090]: Can this change be done in a separate PR? It looks like this change makes it so this bugfix cannot be backported.

**Review by samdoran — COMMENTED** [review]
Changes look good overall. Just one question about using `self.nobest`.

File: lib/ansible/modules/dnf.py (hunk 1 above)
Comment (maxamillion) [reply lib/ansible/modules/dnf.py:1090]: :+1:

**Comment by lerminou** [comment]
Hi, can it be merged ?  @maxamillion

**Comment by maxamillion** [comment]
It has merge conflicts, so no. However, otherwise it looks good. 👍

**Comment by LRitzdorf** [comment]
I'm also affected by this bug, and given the conflicted state of this PR, I'd be willing to rebase it on the current version of Ansible. Before I start, are there any pieces of protocol I should be aware of?

**Comment by s-hertel** [comment]
Merged in #78158. Thanks for the helpful code comments and thorough test cases.
